eng/code-quality-reports/src/main/java/com/azure/tools/checkstyle/checks/ServiceClientInstantiationCheck.java

```diff
@@ -0,0 +1,169 @@
+ *   <li>No public static method named 'builder'</li>
+ *   <li>Since these classes are supposed to be immutable, all fields in the service client classes should be final.</li>
+ * </0l>
+ */
+public class ServiceClientInstantiationCheck extends AbstractCheck {
+    private static final String SERVICE_CLIENT = "ServiceClient";
+    private static final String BUILDER = "builder";
+    private static final String ASYNC_CLIENT ="AsyncClient";
+    private static final String CLIENT = "Client";
+
+    private static boolean hasServiceClientAnnotation;
+    private static boolean notServiceClass;
+
+    @Override
+    public int[] getDefaultTokens() {
+        return getRequiredTokens();
+    }
+
+    @Override
+    public int[] getAcceptableTokens() {
+        return getRequiredTokens();
+    }
+
+    @Override
+    public int[] getRequiredTokens() {
+        return new int[] {
+            TokenTypes.CLASS_DEF,
+            TokenTypes.CTOR_DEF,
+            TokenTypes.METHOD_DEF,
+            TokenTypes.LITERAL_CLASS,
+            TokenTypes.OBJBLOCK
+        };
+    }
+
+    @Override
+    public void beginTree(DetailAST root) {
+        notServiceClass = false;
+    }
+
+    @Override
+    public void visitToken(DetailAST token) {
+        if (notServiceClass) {
+            return;
+        }
+
+        switch (token.getType()) {
+            case TokenTypes.CLASS_DEF:
+                hasServiceClientAnnotation = checkServiceClientAnnotation(token);
+                if (!hasServiceClientAnnotation) {
+                    notServiceClass = true;
+                }
+                break;
+            case TokenTypes.CTOR_DEF:
+                checkNonPublicOrProtectedConstructor(token);
+                break;
+            case TokenTypes.METHOD_DEF:
+                checkNonPublicStaticBuilderMethodName(token);
+                break;
+            case TokenTypes.LITERAL_CLASS:
+                checkServiceClientNaming(token);
+                break;
+            case TokenTypes.OBJBLOCK:
+                checkClassFieldFinal(token);
+                break;
+            default:
+                // Checkstyle complains if there's no default block in switch
+                break;
+        }
+    }
+
+    /**
+     *  Check if the class is annotated with @ServiceClient
+     *
+     * @param token the CLASS_DEF AST node
+     * @return true if the class is annotated with @ServiceClient, false otherwise.
+     */
+    private boolean checkServiceClientAnnotation(DetailAST token) {
+        DetailAST modifiersToken = token.findFirstToken(TokenTypes.MODIFIERS);
+        if (modifiersToken != null) {
+            DetailAST annotationAST = modifiersToken.findFirstToken(TokenTypes.ANNOTATION);
+            if (annotationAST != null) {
+                DetailAST annotationIdent = annotationAST.findFirstToken(TokenTypes.IDENT);
+                if (annotationIdent != null && SERVICE_CLIENT.equals(annotationIdent.getText())) {
```

Comment:
You can return the result of this if statement rather than "return true"

Reply:
I found there is one condition I don't take care of: There could be multiple annotations applied to the same class. The modified method
```
    private boolean hasServiceClientAnnotation(DetailAST token) {
        DetailAST modifiersToken = token.findFirstToken(TokenTypes.MODIFIERS);
        if (modifiersToken != null) {
            for (DetailAST ast = modifiersToken.getFirstChild(); ast != null; ast = ast.getNextSibling()) {
                if (ast.getType() == TokenTypes.ANNOTATION) {
                    DetailAST annotationIdent = ast.findFirstToken(TokenTypes.IDENT);
                    if (annotationIdent != null && SERVICE_CLIENT.equals(annotationIdent.getText())) {
                        return true;
                    }
                }
            }
        }
        return false;
    }
```